### PR TITLE
CDAP-15813: Vertica export plugin validations.

### PR DESCRIPTION
**Pull request by AntonGensitskiy**
JIRA: https://issues.cask.co/browse/CDAP-15813
WIKI: https://wiki.cask.co/display/CE/Guidelines+on+Plugin+Validation

In scope of this PR:

- integrated with new validation API

### CODE REVIEW

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java [hunk 1]
```diff
@@ -160,5 +176,35 @@ public VerticaConfig(String connectionString, String user, String password, Stri
       this.delimiter = delimiter;
       this.path = path;
     }
+
+    public void validate(FailureCollector failureCollector) {
+      if (Strings.isNullOrEmpty(path)) {
+        failureCollector.addFailure("File path is required.",null)
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: provide corrective action. For example "File path must be specified"

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java [hunk 2]
```diff
@@ -160,5 +176,35 @@ public VerticaConfig(String connectionString, String user, String password, Stri
...
+          .withConfigProperty(PATH);
+      }
+      if (Strings.isNullOrEmpty(connectionString)) {
+        failureCollector.addFailure("Connection string is required.", null)
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: provide corrective action. For example "File path must be specified"

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java [hunk 3]
```diff
@@ -160,5 +176,35 @@ public VerticaConfig(String connectionString, String user, String password, Stri
...
+          .withConfigProperty(CONNECTION_STRING);
+      } else if (!connectionString.startsWith(CONNECTION_STRING_PREFIX)) {
+        failureCollector.addFailure(
+          "Connection string has incorrect format.",
+          "Connection string must comply next format - jdbc:vertica://<VerticaHost>:<portNumber>/<databaseName>")
+          .withConfigProperty(CONNECTION_STRING);
+      }
+      if (user == null && password != null) {
+        failureCollector.addFailure(
+          String.format("'%s' is null.", USER),
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: 1. use property name from UI widget, not from constant (in this case just start from uppercase letter)
2. check for empty string (Strings class utility methods)
3. "User is not specified"

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java [hunk 4]
```diff
@@ -160,5 +176,35 @@ public VerticaConfig(String connectionString, String user, String password, Stri
...
+          String.format("Please provide both user name and password if database requires " +
+                          "authentication. If not, please remove '%s' and retry.", USER))
+          .withConfigProperty(USER);
+      }
+      if (user != null && password == null) {
+        failureCollector.addFailure(
+          String.format("'%s' is null.", PASSWORD),
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: 1. use property name from UI widget, not from constant (in this case just start from uppercase letter)
2. check for empty string (Strings class utility methods)
3. "Password is not specified"

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java [hunk 5]
```diff
@@ -104,16 +118,18 @@ public void run(ActionContext context) throws Exception {
   }
 
   /**
-   * Vertica config
+   * Vertica Export config
    */
-  public class VerticaConfig extends PluginConfig {
+  public static class VerticaExportConfig extends PluginConfig {
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: 1. move this class to upper level (make not inner)
2. class should be immutable
3. add builder

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    VerticaBulkExportAction.VerticaExportConfig config = getValidConfig();
+    MockFailureCollector failureCollector = new MockFailureCollector(MOCK_STAGE);
+    config.validate(failureCollector);
+    Assert.assertTrue(failureCollector.getValidationFailures().isEmpty());
+  }
+
+  @Test
+  public void testValidateConnectionStringNull() {
+    VerticaBulkExportAction.VerticaExportConfig config = getValidConfig();
+    config.connectionString = null;
```
Comment (mlozbin-cybervisiontech) [inline src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: use builder, not mutable fields

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    VerticaBulkExportAction.VerticaExportConfig config = getValidConfig();
+    config.password = null;
+    MockFailureCollector failureCollector = new MockFailureCollector(MOCK_STAGE);
+    config.validate(failureCollector);
+    assertValidationFailed(failureCollector,
+                           Collections.singletonList(VerticaBulkExportAction.VerticaExportConfig.PASSWORD), 1);
+  }
+
+  @Test
+  public void testValidateConfig() {
```
Comment (mlozbin-cybervisiontech) [inline src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: this method name does not describe what is tested

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+      "select * from testTable",
+      ",",
+      "/tmp/vertica/vertice_test.csv");
+  }
+
+  private static void assertValidationFailed(MockFailureCollector failureCollector, List<String> paramName,
+                                             int failuresCount) {
+    List<ValidationFailure> failureList = failureCollector.getValidationFailures();
+    Assert.assertEquals(failuresCount, failureList.size());
+    for (int i = 0; i < failureList.size(); i++) {
```
Comment (mlozbin-cybervisiontech) [inline src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: why don't you use foreach? Why do you need to check that failures were collected in same order as params in paramName list?

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+      "jdbc:vertica://localhost:5433/test",
+      "dbadmin",
+      "testpassword",
+      "select * from testTable",
+      ",",
+      "/tmp/vertica/vertice_test.csv");
+  }
+
+  private static void assertValidationFailed(MockFailureCollector failureCollector, List<String> paramName,
+                                             int failuresCount) {
```
Comment (mlozbin-cybervisiontech) [inline src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: in what case failuresCount is not same as paramName.size?

**Review by mlozbin-cybervisiontech — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java (hunk 1 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: Done.

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java (hunk 2 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: Done.

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java (hunk 3 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: Done.

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java (hunk 4 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: Done.

File: src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java (hunk 5 above)
Comment (AntonGensitskiy) [reply src/main/java/io/cdap/plugin/db/batch/action/vertica/export/VerticaBulkExportAction.java]: Done.

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java (hunk 6 above)
Comment (AntonGensitskiy) [reply src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: Done.

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java (hunk 7 above)
Comment (AntonGensitskiy) [reply src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: Name was updated,

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java (hunk 8 above)
Comment (AntonGensitskiy) [reply src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: Method was updated.

File: src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java (hunk 9 above)
Comment (AntonGensitskiy) [reply src/test/java/io/cdap/plugin/db/batch/action/vertice/export/VerticaBulkExportActionConfigTest.java]: Method was updated.

**Pull request merged**